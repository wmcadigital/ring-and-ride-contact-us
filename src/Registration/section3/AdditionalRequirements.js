import { useFormState } from "react-final-form";

import FormSection from "../../common/FormSection";
import RadioGroup from "../../common/RadioGroup";
import Question from "../../common/Question";
import RadioButton from "../../common/RadioButton";
import TextArea from "../../common/TextArea";
import FieldError from "../../common/FieldError";
import ProgressIndicator from "../../common/ProgressIndicator";
import getRequirementsSectionName from "./getRequirementsSectionName";
import { required } from "../../common/validation";

const AdditionalRequirements = () => {
  const stateApi = useFormState();
  const formValues = stateApi.values;

  const error = stateApi.submitFailed
    ? stateApi.errors?.additionalRequirements
    : null;

  const additionalRequirementsDetailsError = stateApi.submitFailed
    ? stateApi.errors?.additionalRequirementsDetails
    : null;

  const question =
    formValues["registerForYourself"] === "yes"
      ? "Do you have any other requirements we need to know about?"
      : `Does ${formValues["firstName"]} have any other requirements we need to know about?`;

  return (
    <FormSection>
      <ProgressIndicator
        sectionPosition="Section 3 of 3"
        sectionName={getRequirementsSectionName(stateApi)}
      />
      <Question text={question} />
      <RadioGroup error={error}>
        <FieldError text={error} />
        <RadioButton
          key={1}
          label="Yes"
          validation={required}
          value="yes"
          fieldName="additionalRequirements"
        />
        {formValues.additionalRequirements === "yes" ? (
          <TextArea
            label="Additional requirements"
            fieldName="additionalRequirementsDetails"
            validation={required}
            error={additionalRequirementsDetailsError}
            rows={8}
          />
        ) : null}
        <RadioButton
          key={2}
          label="No"
          validation={required}
          value="no"
          fieldName="additionalRequirements"
        />
      </RadioGroup>
    </FormSection>
  );
};

export default AdditionalRequirements;
