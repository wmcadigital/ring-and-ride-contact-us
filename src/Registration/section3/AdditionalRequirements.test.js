import { useFormState } from "react-final-form";
import { createRenderer } from "react-test-renderer/shallow";

import AdditionalRequirements from "./AdditionalRequirements";

jest.mock("react-final-form");

describe("AdditionalRequirements - snapshots", () => {
  it("individual is filling out form for themselves", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<AdditionalRequirements />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("individual is filling out form for someone else", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "no",
        firstName: "Test",
        lastName: "Testerson",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<AdditionalRequirements />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an error", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
      },
      submitFailed: true,
      errors: { additionalRequirements: "Required" },
    }));
    const renderer = createRenderer();

    renderer.render(<AdditionalRequirements />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("shows text area for additional requirements if yes is selected", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
        additionalRequirements: "yes",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<AdditionalRequirements />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("shows error if text area not filled in on submission", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
        additionalRequirements: "yes",
      },
      submitFailed: true,
      errors: { additionalRequirementsDetails: "Required" },
    }));
    const renderer = createRenderer();

    renderer.render(<AdditionalRequirements />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
