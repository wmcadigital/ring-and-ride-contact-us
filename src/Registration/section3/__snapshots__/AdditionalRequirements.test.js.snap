// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalRequirements - snapshots individual is filling out form for someone else 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Applicant requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Does Test have any other requirements we need to know about?"
  />
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="additionalRequirements"
      label="Yes"
      validation={[Function]}
      value="yes"
    />
    <RadioButton
      fieldName="additionalRequirements"
      label="No"
      validation={[Function]}
      value="no"
    />
  </RadioGroup>
</FormSection>
`;

exports[`AdditionalRequirements - snapshots individual is filling out form for themselves 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Do you have any other requirements we need to know about?"
  />
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="additionalRequirements"
      label="Yes"
      validation={[Function]}
      value="yes"
    />
    <RadioButton
      fieldName="additionalRequirements"
      label="No"
      validation={[Function]}
      value="no"
    />
  </RadioGroup>
</FormSection>
`;

exports[`AdditionalRequirements - snapshots renders with an error 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Do you have any other requirements we need to know about?"
  />
  <RadioGroup
    error="Required"
  >
    <FieldError
      text="Required"
    />
    <RadioButton
      fieldName="additionalRequirements"
      label="Yes"
      validation={[Function]}
      value="yes"
    />
    <RadioButton
      fieldName="additionalRequirements"
      label="No"
      validation={[Function]}
      value="no"
    />
  </RadioGroup>
</FormSection>
`;

exports[`AdditionalRequirements - snapshots shows error if text area not filled in on submission 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Do you have any other requirements we need to know about?"
  />
  <RadioGroup>
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="additionalRequirements"
      label="Yes"
      validation={[Function]}
      value="yes"
    />
    <TextArea
      error="Required"
      fieldName="additionalRequirementsDetails"
      label="Additional requirements"
      rows={8}
      validation={[Function]}
    />
    <RadioButton
      fieldName="additionalRequirements"
      label="No"
      validation={[Function]}
      value="no"
    />
  </RadioGroup>
</FormSection>
`;

exports[`AdditionalRequirements - snapshots shows text area for additional requirements if yes is selected 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Do you have any other requirements we need to know about?"
  />
  <RadioGroup
    error={null}
  >
    <FieldError
      text={null}
    />
    <RadioButton
      fieldName="additionalRequirements"
      label="Yes"
      validation={[Function]}
      value="yes"
    />
    <TextArea
      error={null}
      fieldName="additionalRequirementsDetails"
      label="Additional requirements"
      rows={8}
      validation={[Function]}
    />
    <RadioButton
      fieldName="additionalRequirements"
      label="No"
      validation={[Function]}
      value="no"
    />
  </RadioGroup>
</FormSection>
`;
